service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpConnectionHandler.java

```diff
@@ -0,0 +1,487 @@
+
+                    case Reject:
+                        log.trace("Amqp message with correlationId {} was rejected, acknowledging it with Rejected", msg.getCorrelationId());
+                        delivery.disposition(new Rejected());
+                        delivery.settle();
+                        break;
+
+                    case Abandon:
+                        log.trace("Amqp message with correlationId {} was abandoned, acknowledging it with Released", msg.getCorrelationId());
+                        delivery.disposition(Released.getInstance());
+                        delivery.settle();
+                        break;
+
+                    default:
+                        log.error("Unknown DeliveryOutcome provided, message will not be acknowledged");
+                }
+            }
+        }
+        else //Sender link received an ack for sent message
+        {
+            log.trace("Received acknowledgement for a message this process sent");
+            Delivery d = event.getDelivery();
+
+            DeliveryState remoteState = d.getRemoteState();
+
+            onMessageAcknowledged(remoteState);
+
+            d.settle();
+        }
+    }
+
+    @Override
+    public void onConnectionBound(Event event)
+    {
+        Transport transport = event.getConnection().getTransport();
+        if (transport != null)
+        {
+            if (this.iotHubServiceClientProtocol == IotHubServiceClientProtocol.AMQPS_WS)
+            {
+                log.debug("Adding websocket layer to transport");
+                WebSocketImpl webSocket = new WebSocketImpl();
+                webSocket.configure(this.webSocketHostName, WEBSOCKET_PATH, 0, WEBSOCKET_SUB_PROTOCOL, null, null);
+                ((TransportInternal)transport).addTransportLayer(webSocket);
+            }
+
+            Sasl sasl = transport.sasl();
+            sasl.plain(this.userName, this.sasToken);
+
+            SslDomain domain = makeDomain(SslDomain.Mode.CLIENT);
+            domain.setPeerAuthentication(SslDomain.VerifyMode.VERIFY_PEER);
+            transport.ssl(domain);
+        }
+    }
+
+    @Override
+    public void onConnectionInit(Event event)
+    {
+        this.connection = event.getConnection();
+        this.connection.setHostname(hostName);
+        Session ssn = this.connection.session();
+
+        Map<Symbol, Object> properties = new HashMap<>();
+        properties.put(Symbol.getSymbol(TransportUtils.versionIdentifierKey), TransportUtils.USER_AGENT_STRING);
+
+        log.trace("Opening connection");
+        this.connection.open();
+
+        log.trace("Opening session");
+        ssn.open();
+        openLinks(ssn, properties);
+    }
+
+    @Override
+    public void onLinkInit(Event event)
+    {
+        Link link = event.getLink();
+        if (event.getLink().getName().equals(tag))
+        {
+            Target t = new Target();
+            t.setAddress(endpoint);
+            Source source = new Source();
```

Comment:
For a c2d link, we don't seem to be setting `Source` currently.
Also, `Target` and `Source` both have the same endpoint; what's the difference?

Reply:
Good question. After digging into this, I found that sender links only need to set the target, and receiver links only need to set the source. I can rewrite this logic to exist only in the AmqpConnectionHandler so that AmqpSendHandler doesn't override this logic